### DISCUSSION TITLE
Using boost and open-mpi with older gcc

### DISCUSSION THREAD

**Discussion by gdolsson** [Homebrew/discussions#506] [Everyday usage]
I am trying to use boost with gcc-9 and I cannot seem to get a hang of how.

I can compile test-code using boost with gcc-10 as g++ though I need to use gcc-9 and in particular with CMake

I managed to figure out how to define using the old gcc packages:

```
-DCMAKE_C_COMPILER=/usr/local/bin/gcc-9 \
-DCMAKE_CXX_COMPILER=/usr/local/bin/g++-9 \
-DCMAKE_Fortran_COMPILER=/usr/local/bin/gfortran-9 \
```

Though the compile breaks at boost and I am trying to figure out how to get boost to work. I tried building the code with macOS clang/clang++ as well though the compile here broke with some problem referencing bison. I suspect that this also has something to do with the fact that I need to use gfortran-9.

Any feedback appreciated
// Gustaf

**Comment by SMillerDev** [comment]
We don't really provide code support. This really sounds like an issue that would fit better at a boost forum or some such.

**Reply by gdolsson** [reply]
Well, yes and no I guess.

What I really wanted to figure out is if, and in that case how to use an older version of homebrew gcc as "the default" gcc version.

The current gcc version is 10 though I need 9. This is not a huge issue as I can specify gcc-9 as mentioned above. The issue I am having is getting the other includes to "follow suite". I also need boost and/or open-mpi, which install perfectly and works with the default gcc (gcc-10) though it will not compile code using gcc-9.

Using MacPorts, there was an option to use "xxx-select" where you could do "gcc-select gcc gcc7" and set gcc7 as the default gcc version, I have read about some deprecated feature called "switch", my question now is if there is a similar option still available for brew?

I need to get around always having to use the most current gcc version, i tried using macOS clang/clang++ though then I need gfortran. Gfortran is now at version 10 together with gcc-10 and will not compile the code. I can specify gfortran-9, though then there is a problem with macOS bison... Trying to manually set compilers to any other than the most current version seems to always cause some problem further down the line and I was curious of there is any way to set a different "default" version of gcc.

Best regards
// Gustaf

**Reply by alebcay** [reply]
> I have read about some deprecated feature called "switch", my question now is if there is a similar option still available for brew?

No, `switch` has been deprecated and there has not been (nor is there, at this time, a plan for) a replacement.

**Reply by gdolsson** [reply]
I keep getting my compilers confused, once again I got turned around. I want to build using GNU gcc compilers and boost. However, the boost I install with brew is built with/for macOS clang/clang++ which is the "default" gcc/g++ compilers on macOS. This is why my sample code works calling g++ to compile the boost including code while it fails calling both g++-10 and g++-9.

Is there any options included in the "brew" ecosystem to solve this? Can I get/compile boost using brew through brew for use with gnu gcc-/g++-9?

**Reply by carlocab** [reply]
Homebrew `boost` is compiled with Clang, but I'm a bit surprised that that makes it useable only with Clang...

If compiling with Clang makes `boost` unusable with `gcc`, then you might need to maintain your own copy of the `boost` formula that's compiled with `gcc@9`. You can do that in your own tap.

See, for example:
https://docs.brew.sh/Taps
https://docs.brew.sh/Interesting-Taps-and-Forks
https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap
https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/
